include/exadg/convection_diffusion/driver.cpp

```diff
@@ -188,28 +187,152 @@ Driver<dim, Number>::ale_update() const
+      {
+        Assert(false, dealii::ExcMessage("ALE and AMR combination not implemented."));
+        pde_operator->initialize_dof_vector_velocity(velocity);
+        pde_operator->interpolate_velocity(velocity, time_integrator->get_time());
+        velocity_ptr = &velocity;
+      }
+
+      auto & grid = *application->get_grid_non_const();
+      grid.coarse_triangulations =
+        dealii::MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence(
+          *application->get_grid_non_const()->triangulation);
+
+      pde_operator->setup_solver(time_integrator_bdf->get_scaling_factor_time_derivative_term(),
+                                 velocity_ptr);
+    }
+    else
+    {
+      AssertThrow(application->get_parameters().temporal_discretization ==
+                    TemporalDiscretization::ExplRK,
+                  dealii::ExcMessage("Not implemented."));
+    }
+  }
+  else
+  {
+    AssertThrow(false, dealii::ExcMessage("Not implemented"));
+  }
+}
+
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::do_adaptive_refinement(unsigned int const time_step_number)
+{
+  dealii::Triangulation<dim> & tria = *application->get_grid_non_const()->triangulation;
+
+  // AMR is only implemented for implicit timestepping.
+  if(application->get_parameters().temporal_discretization != TemporalDiscretization::BDF)
+  {
+    AssertThrow(false, dealii::ExcNotImplemented());
+  }
+  else
+  {
+    std::shared_ptr<TimeIntBDF<dim, Number>> bdf_time_integrator =
+      std::dynamic_pointer_cast<TimeIntBDF<dim, Number>>(time_integrator);
+
+    if(trigger_coarsening_and_refinement_now(application->get_parameters().amr_data,
+                                             time_step_number))
+    {
+      mark_cells_coarsening_and_refinement(tria, bdf_time_integrator->get_solution_np());
+
+      limit_coarsening_and_refinement(tria, application->get_parameters().amr_data);
+
+      if(any_cells_flagged_for_coarsening_or_refinement(tria))
+      {
+        tria.prepare_coarsening_and_refinement();
+
+        time_integrator->prepare_coarsening_and_refinement();
+
+        tria.execute_coarsening_and_refinement();
+
+        setup_after_coarsening_and_refinement();
+
+        time_integrator->interpolate_after_coarsening_and_refinement();
+      }
+    }
+  }
+}
+
 template<int dim, typename Number>
 void
 Driver<dim, Number>::solve()
 {
   if(application->get_parameters().problem_type == ProblemType::Unsteady)
   {
-    if(application->get_parameters().ale_formulation == true)
+    if(application->get_parameters().enable_adaptivity)
     {
       do
       {
         time_integrator->advance_one_timestep_pre_solve(true);
 
-        ale_update();
-
         time_integrator->advance_one_timestep_solve();
 
+        // Adapt the mesh before post_solve(), in order to recalculate the
+        // time step size based on the new mesh.
+        if(time_integrator->get_number_of_time_steps() > 0)
```

Comment:
From a design perspective, I do not see why this if-statement appears here and not further down in the sequence of function calls.

Reply:
I removed it here, since it is not necessary. I did not want to do AMR in the first time step for no reason whatsoever, which was purely arbitrary ... comparing the test results, we see slightly lower errors now doing one more refinement.